java-jsp/src/main/java/org/sonar/java/jsp/Jasper.java

```diff
@@ -33,38 +35,71 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
-import org.apache.jasper.JspC;
+import javax.servlet.jsp.JspFactory;
+import org.apache.jasper.JasperException;
+import org.apache.jasper.JspCompilationContext;
+import org.apache.jasper.compiler.Compiler;
+import org.apache.jasper.compiler.JspRuntimeContext;
+import org.apache.jasper.runtime.JspFactoryImpl;
+import org.apache.jasper.servlet.JspCServletContext;
 import org.sonar.api.batch.fs.FileSystem;
 import org.sonar.api.batch.fs.InputFile;
 import org.sonar.api.batch.sensor.SensorContext;
 import org.sonar.api.scanner.ScannerSide;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
-import org.sonar.java.AnalysisException;
 import org.sonar.java.model.GeneratedFile;
 import org.sonar.java.model.SmapFile;
 
-import static java.util.Arrays.asList;
-
 @ScannerSide
 public class Jasper {
 
   private static final Logger LOG = Loggers.get(Jasper.class);
 
-  public Collection<GeneratedFile> generateFiles(SensorContext context, List<File> javaClasspath) {
-    List<Path> jspFiles = jspFiles(context.fileSystem());
+  public Collection<GeneratedFile> generateFiles(SensorContext sensorContext, List<File> javaClasspath) {
+    List<Path> jspFiles = jspFiles(sensorContext.fileSystem());
     LOG.debug("Found {} JSP files.", jspFiles.size());
     if (jspFiles.isEmpty()) {
       return Collections.emptyList();
     }
-    Path outputDir = outputDir(context);
+    Path outputDir = outputDir(sensorContext);
+    File baseDir = sensorContext.fileSystem().baseDir();
+    // Jasper internally calls Thread#getContextClassLoader to instantiate some classes. ContextClassLoader is set by scanner
+    // and doesn't contain plugin jar, so we need to configure ContextClassLoader with the class loader of the plugin to be able
+    // to run Jasper. Original classloader is restored in finally.
+    ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
     try {
-      Jasper.compileJspFiles(jspFiles, javaClasspath, outputDir, context.fileSystem().baseDir());
+      ClassLoader classLoader = initClassLoader(javaClasspath);
+      Thread.currentThread().setContextClassLoader(classLoader);
+      JspFactory.setDefaultFactory(new JspFactoryImpl());
+      JspCServletContext servletContext = new ServletContext(toUrl(baseDir), classLoader);
+      JasperOptions options = new JasperOptions(servletContext, outputDir);
+      JspRuntimeContext runtimeContext = new JspRuntimeContext(servletContext, options);
+
+      boolean errorTranspiling = false;
+      for (Path file : jspFiles) {
+        LOG.debug("Transpiling JSP: {}", file);
+        try {
+          String relativePath = file.toAbsolutePath().toString().substring(baseDir.getAbsolutePath().length());
```

Comment:
I think this operation you should rely on https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#relativize(java.nio.file.Path) rather than string.